In [1]:
import pandas as pd

# Leer Contenido

In [2]:
datos = pd.read_csv('job_postings.csv')
datos

job_id                                       title             location  \
0          1                            Marketing Intern     US, NY, New York   
1          2   Customer Service - Cloud Video Production       NZ, , Auckland   
2          3     Commissioning Machinery Assistant (CMA)        US, IA, Wever   
3          4           Account Executive - Washington DC   US, DC, Washington   
4          5                         Bill Review Manager   US, FL, Fort Worth   
...      ...                                         ...                  ...   
2860   17828  Student Positions Part-Time and Full-Time.  US, CA, Los Angeles   
2861   17829                             Sales Associate      AU, NSW, Sydney   
2862   17830                           Android Developer       PL, MZ, Warsaw   
2863   17831                              Payroll Clerk      US, NY, New York   
2864   17832                             Furniture mover      US, IL, Chicago   

               department salary_range  \
0               Marketing          NaN   
1                 Success          NaN   
2                     NaN          NaN   
3                   Sales          NaN   
4                     NaN          NaN   
...                   ...          ...   
2860                  NaN          NaN   
2861  Sales and Marketing          NaN   
2862                  NaN          NaN   
2863                  NaN          NaN   
2864                  NaN          NaN   

                                        company_profile  \
0     We're Food52, and we've created a groundbreaki...   
1     90 Seconds, the worlds Cloud Video Production ...   
2     Valor Services provides Workforce Solutions th...   
3     Our passion for improving quality of life thro...   
4     SpotSource Solutions LLC is a Global Human Cap...   
...                                                 ...   
2860                                                NaN   
2861                                                NaN   
2862                                                NaN   
2863                                                NaN   
2864  Anthony Warren is a Marketing and Advertising ...   

                                            description  \
0     Food52, a fast-growing, James Beard Award-winn...   
1     Organised - Focused - Vibrant - Awesome!Do you...   
2     Our client, located in Houston, is actively se...   
3     THE COMPANY: ESRI – Environmental Systems Rese...   
4     JOB TITLE: Itemization Review ManagerLOCATION:...   
...                                                 ...   
2860  Student Positions Part-Time and Full-Time.You ...   
2861  LEARN TO EARN AN EXECUTIVE LEVEL INCOMEFULL TR...   
2862  inFullMobile Sp. z o.o. is a mobile software d...   
2863  JOB DESCRIPTIONWe are seeking a full time payr...   
2864  earn $500 to $1000 a week as a mover.  Must ha...   

                                           requirements  \
0     Experience with content management systems a m...   
1     What we expect from you:Your key responsibilit...   
2     Implement pre-commissioning and commissioning ...   
3     EDUCATION: Bachelor’s or Master’s in GIS, busi...   
4     QUALIFICATIONS:RN license in the State of Texa...   
...                                                 ...   
2860                                                NaN   
2861  What You Can Do. • Have the potential to earn ...   
2862  • A proven track record in Android / JAVA proj...   
2863  JOB REQUIREMENTS•    High school diploma or eq...   
2864           Good liftflexibile schedulegeat attitude   

                                               benefits  telecommuting  \
0                                                   NaN              0   
1     What you will get from usThrough being part of...              0   
2                                                   NaN              0   
3     Our culture is anything but corporate—we have ...              0   
4                              

In [4]:
docs = datos['description'] 
categs = datos['fraudulent'] 

# Obtener e imprimir matriz TFIDF

In [7]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()
docs_bow = vectorizer.fit_transform(docs)

In [8]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf = TfidfTransformer() 
docs_tfidf = tfidf.fit_transform(docs_bow)
docs_tfidf_densa = docs_tfidf.todense()
docs_tfidf_densa

matrix([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.07446071, 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        ...,
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ]])

# Entrenamiento Arbol de decision

In [9]:
from sklearn.model_selection import train_test_split
docs_train, docs_test, categs_train, categs_test = train_test_split(docs_tfidf, categs, test_size = 0.25, 
                                                                    random_state = 50)

In [10]:
from sklearn import tree

a_dec = tree.DecisionTreeClassifier()
a_dec.fit(docs_train, categs_train)

acc_train = a_dec.score(docs_train, categs_train)
acc_test = a_dec.score(docs_test, categs_test)

print("Accuracy entrenamiento: ", acc_train)
print("Accuracy PRUEBA: ", acc_test)
print("Fiabilidad: ", acc_test / acc_train)  

Accuracy entrenamiento:  1.0
Accuracy PRUEBA:  0.8270571827057183
Fiabilidad:  0.8270571827057183


In [11]:

from sklearn.metrics import confusion_matrix
categs_pred = a_dec.predict(docs_test)

cm = confusion_matrix(categs_test, categs_pred)
cm

array([[414,  71],
       [ 53, 179]])

### Podemos ver un accuracy de entrenamiento de 1, que comparandolo con el ACC de prueba, esta gran diferencia de 1 a 0.83, nos señala overfitting, con lo que deberiamos hacer el modelo menos complejo, por ejemplo usando un regularizador para que el modelo generalice mejor.

### Y luego viendo la matriz de confusion podemos ver que tenemos 71 Falsos Positivos y 53 Falsos Negativos, habria que trabajar en bajar estos falsos para obtener una mayor precision de nuestro modelo, seguramente con un regularizador como dicho anteriormente, estos valores bajarian. 

--------